include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1925 @@
+    {
+        std::ignore = n;
+        XTENSOR_ASSERT_MSG(n == size(), "xtiny_impl(n): size mismatch");
+        base_type::fill(v);
+    }
+
+    template <class V, index_t N>
+    inline
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::xtiny_impl(size_type n, tags::skip_initialization_tag)
+    {
+        std::ignore = n;
+        XTENSOR_ASSERT_MSG(n == size(), "xtiny_impl(n): size mismatch.");
+    }
+
+    template <class V, index_t N>
+    template <class IT, class>
+    inline
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::xtiny_impl(IT begin)
+    {
+        assign(begin, begin+N);
+    }
+
+    template <class V, index_t N>
+    template <class IT, class>
+    inline
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::xtiny_impl(IT begin, IT end)
+    {
+        assign(begin, end);
+    }
+
+    template <class V, index_t N>
+    inline
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::xtiny_impl(std::initializer_list<value_type> const & v)
+    {
+        const size_t n = v.size();
+        if(n == 1)
+        {
+            assign(N, static_cast<value_type>(*v.begin()));
+        }
+        else if(n == N)
+        {
+            assign(v.begin(), v.end());
+        }
+        else
+        {
+            XTENSOR_ASSERT_MSG(false, "xtiny_impl::xtiny_impl(std::initializer_list): size mismatch.");
+        }
+    }
+
+    template <class V, index_t N>
+    inline
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::xtiny_impl(xtiny_impl const & v)
+    : base_type(v)
+    {
+    }
+
+    template <class V, index_t N>
+    inline
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::xtiny_impl(xtiny_impl && v)
+    : base_type(std::forward<base_type>(v))
+    {
+    }
+
+    template <class V, index_t N>
+    inline auto
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::operator=(xtiny_impl const & v) -> xtiny_impl &
+    {
+        base_type::operator=(v);
+        return *this;
+    }
+
+    template <class V, index_t N>
+    inline auto
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::operator=(xtiny_impl && v) -> xtiny_impl &
+    {
+        base_type::operator=(std::forward<base_type>(v));
+        return *this;
+    }
+
+    template <class V, index_t N>
+    inline void
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::assign(size_type n, const value_type& v)
+    {
+        std::ignore = n;
+        XTENSOR_ASSERT_MSG(n == size(), "xtiny_impl::assign(n, v): size mismatch.");
+        base_type::fill(v);
+    }
+
+    template <class V, index_t N>
+    template <class IT, class>
+    inline void
+    xtiny_impl<V, N, std::array<V, (size_t)N>>::assign(IT begin, IT end)
+    {
+        std::ignore = end;
+        XTENSOR_ASSERT_MSG(std::distance(begin, end) == static_cast<std::ptrdiff_t>(size()),
+            "xtiny_impl::assign(begin, end): size mismatch.");
+        for(size_type k=0; k<N; ++k, ++begin)
+        {
+            (*this)[k] = static_cast<value_type>(*begin);
```

Comment:
i think we should use iterators here, too. 
And I am unsure about the static_cast. Is that correct behavior compared with e.g. std::vector?

Reply:
> i think we should use iterators here, too.

I don't mind as long as the compiler manages to unroll these loops (or applies equally powerful optimization) when `N` is known at compile time. 

> And I am unsure about the static_cast. Is that correct behavior compared with e.g. std::vector?

In principle, `std::copy` could be called here, but it issues warnings when `decltype(*begin)` doesn't match `value_type`. Since `xtiny` is meant to be usable beyond shapes (e.g. for RGB pixels), mixed-type expressions are not rare.

The warning policies of current compilers w.r.t. automatic type conversion amount to the implicit rule "automatic type conversion is deprecated". From the image processing point of view, I don't agree with this rule, and add explicit casts where I think automatic conversion should be allowed.